handle/reformatEtcdPath.go

```diff
@@ -0,0 +1,191 @@
+package handle
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/pingcap/log"
+	"github.com/ystaticy/serverless_keyspace_tools/common"
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/client/v3/concurrency"
+	"go.uber.org/zap"
+)
+
+const (
+	pathPrefix   = "/keyspaces/tidb/"
+	reformatLock = "/keyspace_reformat_lock"
+	target       = "//"
+	replacement  = "/"
+)
+
+type (
+	task   []*mvccpb.KeyValue
+	result int
+)
+
+var (
+	cli *clientv3.Client
+	mu  *concurrency.Mutex
+)
+
+func ReformatEtcdPath(ctx context.Context, pdAddr string, workerCount int, targetKeyspaceID string, run bool) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var err error
+	// Connect to etcd.
+	cli, err = clientv3.New(clientv3.Config{
+		Endpoints: []string{pdAddr},
+	})
+	if err != nil {
+		return err
+	}
+
+	// Get all keys that needs to be fixed and batch them into jobs.
+	var res *clientv3.GetResponse
+	if len(targetKeyspaceID) != 0 {
+		res, err = cli.Get(ctx, pathPrefix+targetKeyspaceID+target, clientv3.WithPrefix())
+	} else {
+		res, err = cli.Get(ctx, pathPrefix, clientv3.WithPrefix())
+	}
+
+	if err != nil {
+		return err
+	}
+
+	var (
+		keysToReformat int
+		skippedKey     int
+	)
+	// A map used to group keyspaces with same ID to the group.
+	jobs := make(map[string]task)
+	for _, kv := range res.Kvs {
+		before, _, exist := strings.Cut(string(kv.Key), target)
+		// Skip keys that do not contain target substring.
+		if !exist {
+			skippedKey++
+			continue
+		}
+		keyspaceID := strings.Trim(before, pathPrefix)
+		keysToReformat++
+		jobs[keyspaceID] = append(jobs[keyspaceID], kv)
+	}
+
+	log.Info("Start to reformat",
+		zap.Int("total scanned key", len(res.Kvs)),
+		zap.Int("keys to reformat", keysToReformat),
+		zap.Int("skipped key", skippedKey),
+		zap.Int("keyspaces to reformat", len(jobs)),
+	)
+
+	for k, v := range jobs {
+		fmt.Printf("-------------------- keyspaceID: %s --------------------\n", k)
+		for _, path := range v {
+			fmt.Printf("\tkey: %s\n\tvalue: %s\n\tlease: %d\n", string(path.Key), string(path.Value), path.Lease)
+		}
+	}
+	if !run {
+		log.Info("not operating in run mode, reformat skipped")
+		return nil
+	}
+
+	if len(jobs) == 0 {
+		log.Info("no keys to reformat, reformatting skipped")
+		return nil
+	}
+
+	// Create mutex for concurrency control.
+	session, err := concurrency.NewSession(cli, concurrency.WithContext(ctx))
+	if err != nil {
+		return err
+	}
+	defer session.Close()
+	mu = concurrency.NewMutex(session, reformatLock)
```

Comment:
Why should we use an `etcd` lock here, seems multiple reformat tasks are allowed to run simultaneously.